---
title: "Estimating diagnostic noise in panel-based genomic analysis"
author: 
  - Robin N. Beaumont:
      institute: exeter
      email: R.Beaumont@exeter.ac.uk
  - Caroline F. Wright:
      institute: exeter
      email: Caroline.Wright@exeter.ac.uk
institute:
  - exeter: Institute of Biomedical and Clinical Science, University of Exeter Medical School, University of Exeter, Exeter EX2 5DW, UK
output:
  bookdown::word_document2:
    fig_caption: yes  # for cross-references
    number_sections: TRUE
    pandoc_args:
      - '--lua-filter=scholarly-metadata.lua'
      - '--lua-filter=author-info-blocks.lua'
params:
  phenos: ""	# location of DDD phenotype data
csl: biomed_central.csl
---

```{r init_lib, message=F, echo=F, results="hide", warning=F}
# Load libraries
library(plyr)
library(tidyverse)
library(magrittr)
library(ggplot2)
library(knitr)
library(grid)
library(gridExtra)
library(gtable)
library(ggpubr)
library(Cairo)
library(ggbreak)
library(patchwork)
# define colour blind palette and useful functions
cbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#000000")  # colour blind palette
word_comment <- function(comment, highlight = "") {
  if (isTRUE(knitr:::pandoc_to() == "docx")) {
      paste0('[', comment, ']{.comment-start id="0" author="Rob"',
                'date="',Sys.time(),'"}', highlight, '[]{.comment-end id="0"}')
  }else{
      paste0(highlight)
  }
}
# read in the data
## counts per gene
ddd<-read_tsv("results/DDD_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
skin<-read_tsv("results/Skin_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
cancer<-read_tsv("results/Cancer_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
eye<-read_tsv("results/Eye_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
ddd_rdif<-read_tsv("results/DDD_RD_IF_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
cardiac<-read_tsv("results/Cardiac_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
ddd_merged<-rbind(ddd,ddd_rdif) %>% mutate(id=paste(gene_name,inheritance,sep="-"))	# merge the 2 DDD panels, then remove duplicate rows generated by this
d2<-ddd_merged %>% select(gene_name,gene_id,transcript_id,inheritance,id)
ddd_merged<-aggregate(ddd_merged[,5:8],list(ddd_merged$id),mean) %>% dplyr::rename(id=Group.1) %>% merge(.,d2) %>% select(gene_name,gene_id,transcript_id,inheritance,variants,individuals,length,total_length)
rm(d2)
### all transcripts
ddd_all<-read_tsv("results/DDD_all_transcripts_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
ddd_rdif_all<-read_tsv("results/DDD_RD_IF_all_transcripts_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
ddd_all_merged<-rbind(ddd_all,ddd_rdif_all) %>% mutate(id=paste(gene_name,inheritance,sep="-"))
d2<-ddd_all_merged %>% select(gene_name,gene_id,transcript_id,inheritance,id)
ddd_all_merged<-aggregate(ddd_all_merged[,5:8],list(ddd_all_merged$id),mean) %>% dplyr::rename(id=Group.1) %>% merge(.,d2) %>% select(gene_name,gene_id,transcript_id,inheritance,variants,individuals,length,total_length)
skin_all<-read_tsv("results/Skin_all_transcripts_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
cancer_all<-read_tsv("results/Cancer_all_transcripts_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
eye_all<-read_tsv("results/Eye_all_transcripts_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
cardiac_all<-read_tsv("results/Cardiac_all_transcripts_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
## counts per individual
ddd_ind<-read_tsv("results/DDD_individual_counts_chrs_collated.counts",show_col_types = FALSE)
ddd_rdif_ind<-read_tsv("results/DDD_RD_IF_individual_counts_chrs_collated.counts",show_col_types = FALSE)     # read in data by individual
ddd_merged_ind<-ddply(merge(ddd_ind,ddd_rdif_ind,all.x=TRUE),.(ID),summarise,count_variants_unique=sum(count_variants_unique),count_variants=sum(count_variants),count_genes=sum(count_genes),count_homs=sum(count_homs),count_monoallelic=sum(count_monoallelic),count_biallelic=sum(count_biallelic))
skin_ind<-read_tsv("results/Skin_individual_counts_chrs_collated.counts",show_col_types = FALSE)
cancer_ind<-read_tsv("results/Cancer_individual_counts_chrs_collated.counts",show_col_types = FALSE)
eye_ind<-read_tsv("results/Eye_individual_counts_chrs_collated.counts",show_col_types = FALSE)
cardiac_ind<-read_tsv("results/Cardiac_individual_counts_chrs_collated.counts",show_col_types = FALSE)
# combine the panels
## gene level
combined_panels<-ddd_merged %>% mutate(panel="DD")
combined_panels<-skin %>% mutate(panel="Skin") %>% rbind(combined_panels,.)
combined_panels<-eye %>% mutate(panel="Eye") %>% rbind(combined_panels,.)
combined_panels<-cancer %>% mutate(panel="Cancer") %>% rbind(combined_panels,.)
combined_panels<-cardiac %>% mutate(panel="Cardiac") %>% rbind(combined_panels,.)
combined_panels_monoallelic<-combined_panels %>% filter(inheritance=="monoallelic")
combined_panels_biallelic<-combined_panels %>% filter(inheritance=="biallelic")
mod_mono<-combined_panels_monoallelic %$% lm(individuals~length)
mod_bi<-combined_panels_biallelic %$% lm(individuals~poly(length,2,raw=T)) # generate model with length^2
# count number of genes in each panel
ddd_len<-nrow(ddd_merged)
ddd_tot<-ddd_merged %$% sum(length)
skin_len<-nrow(skin)
skin_tot<-skin %$% sum(length)
cancer_len<-nrow(cancer)
cancer_tot<-cancer %$% sum(length)
eye_len<-nrow(eye)
eye_tot<-eye %$% sum(length)
cardiac_len<-nrow(cardiac)
cardiac_tot<-cardiac %$% sum(length)
## Read in data for ClinVar and REVEL filters
ddd_clinvar_revel<-read_tsv("results/DDD_individual_counts_chrs_collated_clinvar_revel.counts",show_col_types = FALSE)
skin_clinvar_revel<-read_tsv("results/Skin_individual_counts_chrs_collated_clinvar_revel.counts",show_col_types = FALSE)
eye_clinvar_revel<-read_tsv("results/Eye_individual_counts_chrs_collated_clinvar_revel.counts",show_col_types = FALSE)
cancer_clinvar_revel<-read_tsv("results/Cancer_individual_counts_chrs_collated_clinvar_revel.counts",show_col_types = FALSE)
cardiac_clinvar_revel<-read_tsv("results/Cardiac_individual_counts_chrs_collated_clinvar_revel.counts",show_col_types = FALSE)
ddd_rdif_clinvar_revel<-read_tsv("results/DDD_RD_IF_individual_counts_chrs_collated_clinvar_revel.counts",show_col_types = FALSE)
ddd_merged_clinvar_revel<-ddply(merge(ddd_clinvar_revel,ddd_rdif_clinvar_revel,all.x=TRUE),.(ID),summarise,count_variants_unique=sum(count_variants_unique),count_variants=sum(count_variants),count_genes=sum(count_genes),count_homs=sum(count_homs),conflicting=sum(conflicting),benign=sum(benign),pathogenic=sum(pathogenic),path_and_benign=sum(path_and_benign),uncertain=sum(uncertain))
# combine and remove the panels
combined_panels_clinvar_revel<-ddd_merged_clinvar_revel %>% mutate(panel="DD",ngene=ddd_len)
combined_panels_clinvar_revel<-skin_clinvar_revel %>% mutate(panel="Skin",ngene=skin_len) %>% rbind(combined_panels_clinvar_revel,.)
combined_panels_clinvar_revel<-eye_clinvar_revel %>% mutate(panel="Eye",ngene=eye_len) %>% rbind(combined_panels_clinvar_revel,.)
combined_panels_clinvar_revel<-cancer_clinvar_revel %>% mutate(panel="Cancer",ngene=cancer_len) %>% rbind(combined_panels_clinvar_revel,.)
combined_panels_clinvar_revel<-cardiac_clinvar_revel %>% mutate(panel="Cardiac",ngene=cardiac_len) %>% rbind(combined_panels_clinvar_revel,.)
rm(ddd_clinvar_revel,skin_clinvar_revel,eye_clinvar_revel,cancer_clinvar_revel,cardiac_clinvar_revel,ddd_rdif_clinvar_revel,ddd_merged_clinvar_revel)

## Cancer cases and controls
cancer_cases<-read_tsv("results/Cancer_cases_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
cancer_controls<-read_tsv("results/Cancer_controls_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
ncase<-read_tsv("cancer_cases",show_col_types = FALSE) %>% nrow()
ncont<-read_tsv("cancer_controls",show_col_types = FALSE) %>% nrow()
res<-ks.test(cancer_cases$individuals/ncase,cancer_controls$individuals/ncont)
violin<-ddd_merged_ind %>% select(-c(count_monoallelic,count_biallelic)) %>% mutate(list="DD",ngene=ddd_len,total_len=ddd_tot)
violin<-skin_ind %>% mutate(list="Skin",ngene=skin_len,total_len=skin_tot) %>% rbind(violin,.)
violin<-cancer_ind %>% mutate(list="Cancer",ngene=cancer_len,total_len=cancer_tot) %>% rbind(violin,.)
violin<-eye_ind %>% mutate(list="Eye",ngene=eye_len,total_len=eye_tot) %>% rbind(violin,.)
violin<-cardiac_ind %>% mutate(list="Cardiac",ngene=cardiac_len,total_len=cardiac_tot) %>% rbind(violin,.)
violin %<>% mutate(list=as.factor(list)) %>% filter(ID>0)
# create dataframes of EUR and non EUR individuals for summary
violin_ddd <- ddd_merged_ind %>% select(-c(count_monoallelic,count_biallelic)) %>% mutate(list="UKB",n=nrow(.))
non_EUR_ind<-read_tsv("UKBB_non_eur_200k",show_col_types = FALSE)
names(non_EUR_ind)<-c("ID")
violin_ddd <- ddd_merged_ind %>% select(-c(count_monoallelic,count_biallelic)) %>% filter(!(ID %in% non_EUR_ind$ID) & ID>0) %>% mutate(list="UKB EUR",n=nrow(.)) %>% rbind(violin_ddd,.)
violin_ddd <- ddd_merged_ind %>% select(-c(count_monoallelic,count_biallelic)) %>% filter(ID %in% non_EUR_ind$ID) %>% mutate(list="UKB non-EUR",n=nrow(.)) %>% rbind(violin_ddd,.)
violin_ddd <- ddd_merged_ind %>% select(ID,count_monoallelic,count_variants,count_genes,count_homs) %>% rename(count_variants_unique=count_monoallelic) %>% mutate(list="UKB monoallelic genes",n=nrow(.)) %>% rbind(violin_ddd,.)
violin_ddd <- ddd_merged_ind %>% select(ID,count_biallelic,count_variants,count_genes,count_homs) %>% rename(count_variants_unique=count_biallelic) %>% mutate(list="UKB biallelic genes",n=nrow(.)) %>% rbind(violin_ddd,.)
# UKBB missense lof
ddd_ukbb_mis<-read_tsv("results/DDD_individual_counts_chrs_collated_missense_lof.counts",show_col_types = FALSE)
# DDD cohort
ddd_cohort_ind<-read_tsv("DDD_cohort/results/DDD_indivudals_variants.counts",show_col_types = FALSE)	# probands
ddd_cohort_parents_ind<-read_tsv("DDD_cohort/results/DDD_indivudals_variants_parents.counts",show_col_types = FALSE)	# parents
# subset to only unaffected parents
phenos<-read_tsv(params$phenos) %>% filter(dad_id==0 & mum_id==0 & affected==1)
# get the IDs of parents who are unaffected
ddd_cohort_parents_ind %<>% mutate(affected=ifelse(ID %in% c(phenos$individual_id),0,1))

ddd_cohort_ind_mis<-read_tsv("DDD_cohort/results/DDD_indivudals_variants_missense_lof.counts",show_col_types = FALSE)	# missense and lof split
ddd_ancestry<-read_tsv("DDD_cohort/ddd.pca_classification_withminimalIDs.txt",show_col_types = FALSE) %>% filter(PCA_ancestry_classification %in% c("EUR"))
violin_ddd <- ddd_cohort_ind %>% select(ID,count_variants_unique,count_variants,count_genes,count_homs) %>% mutate(list="DDD probands",n=nrow(.)) %>% rbind(violin_ddd,.)
violin_ddd <- ddd_cohort_parents_ind %>% mutate(list="DDD parents",n=nrow(.)) %>% select(-affected) %>% rbind(violin_ddd,.)
violin_ddd <- ddd_cohort_parents_ind %>% filter(affected==0) %>% mutate(list="DDD parents (unaffected)",n=nrow(.)) %>% select(-affected) %>% rbind(violin_ddd,.)
violin_ddd <- ddd_cohort_parents_ind %>% filter(affected==1) %>% mutate(list="DDD parents (affected)",n=nrow(.)) %>% select(-affected) %>% rbind(violin_ddd,.)
violin_ddd <- ddd_cohort_ind %>% select(ID,count_variants_unique,count_variants,count_genes,count_homs) %>% filter(ID %in% ddd_ancestry$person_stable_id) %>% mutate(list="DDD probands EUR",n=nrow(.)) %>% rbind(violin_ddd,.)
violin_ddd <- ddd_cohort_ind %>% select(ID,count_variants_unique,count_variants,count_genes,count_homs) %>% filter(!(ID %in% ddd_ancestry$person_stable_id)) %>% mutate(list="DDD probands non-EUR",n=nrow(.)) %>% rbind(violin_ddd,.)
violin_ddd <- ddd_cohort_ind %>% select(ID,count_monoallelic,count_variants,count_genes,count_homs) %>% rename(count_variants_unique=count_monoallelic) %>% mutate(list="DDD probands monoallelic genes",n=nrow(.)) %>% rbind(violin_ddd,.)
violin_ddd <- ddd_cohort_ind %>% select(ID,count_biallelic,count_variants,count_genes,count_homs) %>% rename(count_variants_unique=count_biallelic) %>% mutate(list="DDD probands biallelic genes",n=nrow(.)) %>% rbind(violin_ddd,.)
violin_ddd <- ddd_ukbb_mis %>% select(ID,lof,count_variants,count_genes,count_homs) %>% relocate(lof,.after=ID) %>% dplyr::rename(count_variants_unique=lof) %>% mutate(list="UKB LoF",n=nrow(.)) %>% rbind(violin_ddd,.)
violin_ddd <- ddd_ukbb_mis %>% select(ID,missense,count_variants,count_genes,count_homs) %>% relocate(missense,.after=ID) %>% dplyr::rename(count_variants_unique=missense) %>% mutate(list="UKB missense",n=nrow(.)) %>% rbind(violin_ddd,.)
# Add in DDD cis and trans
ddd_cis <- read_tsv("DDD_cohort/results/DDD_all_chr_variant_counts_biallelic_cis_trans.counts")
cis_tmp <- ddd_cis %>% select(c(names(ddd_cis)[grepl("cis",names(ddd_cis))])) %>% rowSums()
trans_tmp <- ddd_cis %>% select(c(names(ddd_cis)[grepl("trans",names(ddd_cis))])) %>% rowSums()
ddd_cis_tmp <- ddd_cis %>% select(c(names(ddd_cis)[grepl("cis",names(ddd_cis))])) # count the number of genes with variants in cis/trans instead of individual variants
for (i in 1:ncol(ddd_cis_tmp)){ddd_cis_tmp[,i]<-ddd_cis_tmp[,i]/ddd_cis_tmp[,i]} %>% rowSums(na.rm=T)
cis_genes <- rowSums(ddd_cis_tmp,na.rm=T)
ddd_cis_tmp <- ddd_cis %>% select(c(names(ddd_cis)[grepl("trans",names(ddd_cis))]))
for (i in 1:ncol(ddd_cis_tmp)){ddd_cis_tmp[,i]<-ddd_cis_tmp[,i]/ddd_cis_tmp[,i]} %>% rowSums(na.rm=T)
trans_genes <- rowSums(ddd_cis_tmp,na.rm=T)
rm(ddd_cis_tmp)
violin_ddd <- ddd_cis %>% select(ID) %>% mutate(count_variants_unique=cis_tmp,count_variants=cis_tmp,count_genes=0,count_homs=0,list="DDD (variants in cis in biallelic genes)",n=nrow(ddd_cohort_ind)) %>% rbind(violin_ddd,.)
violin_ddd <- ddd_cis %>% select(ID) %>% mutate(count_variants_unique=trans_tmp,count_variants=trans_tmp,count_genes=0,count_homs=0,list="DDD (variants in trans in biallelic genes)",n=nrow(ddd_cohort_ind)) %>% rbind(violin_ddd,.)
rm(cis_tmp,trans_tmp,ddd_cis)
ddd_cohort_ind_mis<-read_tsv("DDD_cohort/results/DDD_indivudals_variants_missense_lof.counts",show_col_types = FALSE)
violin_ddd<-ddd_cohort_ind_mis %>% select(ID,lof,count_variants,count_genes,count_homs) %>% relocate(lof, .after=ID) %>% dplyr::rename(count_variants_unique=lof) %>% mutate(list="DDD probands LoF",n=nrow(.)) %>% rbind(violin_ddd,.)
violin_ddd<-ddd_cohort_ind_mis %>% select(ID,missense,count_variants,count_genes,count_homs) %>% relocate(missense, .after=ID) %>% dplyr::rename(count_variants_unique=missense) %>% mutate(list="DDD probands missense",n=nrow(.)) %>% rbind(violin_ddd,.)
violin_ddd %<>% mutate(list=factor(list,levels=c("UKB","UKB EUR","UKB non-EUR","UKB LoF","UKB missense","UKB monoallelic genes","UKB biallelic genes","DDD parents","DDD parents (unaffected)","DDD parents (affected)","DDD probands","DDD probands EUR","DDD probands non-EUR","DDD probands LoF","DDD probands missense","DDD probands monoallelic genes","DDD probands biallelic genes","DDD (variants in cis in biallelic genes)","DDD (variants in trans in biallelic genes)")))
```

\newpage

```{r table1,echo=F, message=F}
table<-data.frame()
# generate table of panels:
# number of genes (combined and split by mon and biallelic)
# genomic footprint (both length and proportion of Exome)
generate_table<-function(data,name,table){
    ngene<-data %$% n_distinct(gene_name)
    n_bi<-data %>% filter(inheritance=="biallelic") %$% n_distinct(gene_name)
    n_mono<-data %>% filter(inheritance=="monoallelic") %$% n_distinct(gene_name)
    glength<-data %>% select(gene_name,length) %>% distinct() %$% sum(length,na.rm=T)
    gprop<-round(glength/592815.18,1) %>% paste0(.,"%")        # length from https://bmcresnotes.biomedcentral.com/articles/10.1186/s13104-019-4343-8
    glength %<>% format(.,big.mark=",",trim=T)
    table<-rbind(table,c(name,ngene,n_bi,n_mono,glength,gprop))
    names(table)<-c("Panel","Number of Genes","Biallelic Genes","Monoallelic Genes","Total length (bp)","Percentage of Exome")
    table
}
table %<>% generate_table(ddd_merged,"DD",.)
table %<>% generate_table(skin,"Skin",.)
table %<>% generate_table(cancer,"Cancer",.)
table %<>% generate_table(eye,"Eye",.)
table %<>% generate_table(cardiac,"Cardiac",.)

# generate table with mean, median etc.
ntot<-nrow(ddd_ind)
tbl_nnt<-violin %>% dplyr::group_by(list) %>% dplyr::summarise(mean_mono=paste0(round(mean(count_variants_unique),1)," (",min(count_variants_unique),"-",max(count_variants_unique),")"))
names(tbl_nnt)<-c("Panel","Mean (minimum-maximum) number of rare variants prioritised in UKB")

# add in number of genes with 0 individuals
table_0<-data.frame()
ddd0<-ddd_all_merged %>% filter(individuals==0) %>% nrow()
table_0<-rbind(table_0,c("DD",ddd0))
skin0<-skin_all %>% filter(individuals==0) %>% nrow()
table_0<-rbind(table_0,c("Skin",skin0))
cancer0<-cancer_all %>% filter(individuals==0) %>% nrow()
table_0<-rbind(table_0,c("Cancer",cancer0))
eye0<-eye_all %>% filter(individuals==0) %>% nrow()
table_0<-rbind(table_0,c("Eye",eye0))
cardiac0<-cardiac_all %>% filter(individuals==0) %>% nrow()
table_0<-rbind(table_0,c("Cardiac",cardiac0))
names(table_0)<-c("Panel","Genes with zero UKB individuals")
merge(table,tbl_nnt) %>% merge(.,table_0) %>% kable(caption="Number of genes in each gene panel, plus mean and range of variants prioritised per person for each gene list.",align=c('l',rep('r', 7)))
```

\newpage

```{r outlyingGenes,warning=F,message=F,echo=F}
mod<-combined_panels_monoallelic %$% lm(individuals~length)
combined_panels_monoallelic %<>% mutate(resid=residuals(mod)+mod$coeff[1],resid2=(residuals(mod)+mod$coeff[1])^2) %>% mutate(outlier=as.factor(ifelse(resid2>mean(resid2)+sd(resid2)*3,1,0)))
combined_panels_monoallelic$gene_name[which(combined_panels_monoallelic$gene_name=="KMT2C,MLL3")]="KMT2C"
combined_panels_monoallelic$gene_name[which(combined_panels_monoallelic$gene_name=="DNCH1,DNCL,DNECL,DYNC1H1")]="DYNC1H1"
combined_panels_monoallelic$gene_name[which(combined_panels_monoallelic$gene_name=="EBS1,PLEC,PLEC1")]="PLEC"
combined_panels_monoallelic$gene_name[which(combined_panels_monoallelic$gene_name=="CMPD4FLJ32040TMDCMH9LGMD2JMYLK5,TTN")]="TTN"
combined_panels_monoallelic$gene_name[which(combined_panels_monoallelic$gene_name=="KMT2B,KMT2D,MLL2,TNRC21")]="KMT2D"
combined_panels_monoallelic$gene_name[which(combined_panels_monoallelic$gene_name=="ARVC2,ARVD2,RYR2,VTSIP")]="RYR2"
combined_panels_monoallelic$gene_name[which(combined_panels_monoallelic$gene_name=="COL7A1,EBD1,EBDCT,EBR1")]="COL7A1"
mod<-combined_panels_biallelic %$% lm(individuals~poly(length,2,raw=T)) # generate model with length^2
combined_panels_biallelic %<>% mutate(resid=individuals-(mod$coeff[1]+mod$coeff[2]*length+mod$coeff[3]*length^2)) %>% mutate(resid2=resid^2,outlier=as.factor(ifelse(resid2>mean(resid2)+sd(resid2)*3,1,0)))    # generate residuals, MSE and outlier flag
combined_panels_biallelic$gene_name[which(combined_panels_biallelic$gene_name=="COCA2,MLH1")]="MLH1"
combined_panels_biallelic$gene_name[which(combined_panels_biallelic$gene_name=="HSPG2,SJS1")]="HSPG2"
combined_panels_biallelic$gene_name[which(combined_panels_biallelic$gene_name=="FAM38A,PIEZO1")]="PIEZO1"
combined_panels_biallelic$gene_name[which(combined_panels_biallelic$gene_name=="C10orf61,TCTN3")]="TCTN3"
combined_panels_biallelic$gene_name[which(combined_panels_biallelic$gene_name=="ADGRV1,GPR98,MASS1,USH2C")]="ADGRV1"
combined_panels_biallelic$gene_name[which(combined_panels_biallelic$gene_name=="EBS1,PLEC,PLEC1")]="PLEC"
combined_panels_biallelic$gene_name[which(combined_panels_biallelic$gene_name=="C6orf98,SYNE1")]="SYNE1"
combined_panels_biallelic$gene_name[which(combined_panels_biallelic$gene_name=="NEB,NEM2")]="NEB"
combined_panels_biallelic$gene_name[which(combined_panels_biallelic$gene_name=="CKN2,ERCC6")]="ERCC6"
combined_panels_biallelic$gene_name[which(combined_panels_biallelic$gene_name=="ALUNC,HR")]="HR"
table_out<-combined_panels_monoallelic %>% filter(outlier==1) %>% select(gene_name,individuals,length,panel) %>% arrange(panel,desc(length)) %>% mutate(type="Monoallelic")
table_out<-combined_panels_biallelic %>% filter(outlier==1) %>% mutate(panel=ifelse(gene_name=="HR","DDG, Skin",ifelse(gene_name=="ERCC6","DD, Eye, Skin",panel))) %>% distinct() %>% select(gene_name,individuals,length,panel) %>% arrange(panel,desc(length)) %>% mutate(type="Biallelic") %>% rbind(table_out,.) %>% mutate(reason="Distance")
```
```{r outlyingGenesSize,warning=F,message=F,echo=F}
table_out<-combined_panels_monoallelic %>% filter(outlier==0 & length>12500) %>% select(gene_name,individuals,length,panel) %>% arrange(panel,desc(length)) %>% mutate(type="Monoallelic") %>% mutate(reason="Gene size") %>% rbind(table_out,.)
table_out<-combined_panels_biallelic %>% filter(outlier==0 & length>20000) %>% select(gene_name,individuals,length,panel) %>% arrange(panel,desc(length)) %>% mutate(type="Biallelic") %>% mutate(reason="Gene size") %>% rbind(table_out,.)
names(table_out)=c("Gene","Individuals","Length (bp)","Panel(s)","Type","Reason")
table_out %>% kable(caption="Outlying genes.")
```

\newpage

```{r compareUKBBtoDDD,echo=F,message=F,warning=F}
# table for comparison of number of variants in DDD vs UKBB
# build the table
tbl_nnt<-violin_ddd %>% dplyr::group_by(list) %>% dplyr::summarise(mean_mono=round(mean(count_variants_unique),1),sd_mono=round(sd(count_variants_unique),1),median_mono=median(count_variants_unique),min_mono=paste0("(",min(count_variants_unique)," - ",max(count_variants_unique),")"),n=mean(n))
names(tbl_nnt)<-c("Cohort","Mean","SD","Median","Minimum - Maximum","N")
kable(tbl_nnt,caption="Subsets of mean, median, minimum, and maximum variants prioritised in the DD panel per individual in the DDD cohort compared with UKB. (EUR = European, LoF = loss-of-function).")
```

\newpage

```{r echo=F,warning=F,message=F,results='hide',fig.show='hide'}
mod<-lm(count_variants~ngene-1,violin)
p1<-ggplot(violin,aes(x=ngene,y=count_variants,group=list,colour=list))+geom_boxplot(outlier.size=0.5)+geom_abline(slope=mod$coeff[1],intercept=0,linetype="dotted")+theme_bw()+xlab("Number of Genes in List")+ylab("Number of Variants")+xlim(0,max(violin$ngene)+100)+scale_y_continuous(breaks=c(0,2,4,6,8,10))+coord_cartesian(xlim=c(-10,max(violin$ngene)+100),ylim=c(-0.1,10.5),expand=F)+theme(legend.title=element_blank())+scale_colour_manual(values=cbPalette)	# create partial plot
p2<-ggplot(violin,aes(x=ngene,y=count_variants,group=list,colour=list))+geom_boxplot(outlier.size=0.5)+geom_abline(slope=mod$coeff[1],intercept=0,linetype="dotted")+theme_bw()+xlab("")+ylab("")+xlim(0,max(violin$ngene)+100)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+scale_y_continuous(breaks=c(20,40,60))+coord_cartesian(xlim=c(-10,max(violin$ngene)+100),ylim=c(11,max(violin$count_variants)+5),expand=F)+theme(legend.position="None")+scale_colour_manual(values=cbPalette)	# create partial plot
legend<-get_legend(p1)
p1<-p1+theme(legend.position="None")
layout <- "
AAAAAAC
BBBBBBC
BBBBBBC
BBBBBBC
"
CairoPNG("figBoxLine.png",width=800,height=400)
p2+p1+legend+plot_layout(design = layout)
dev.off()
```
```{r box,fig.cap="Number of rare non-synonymous variants found in each person in UKB for each gene panel vs number of genes in panel.",warning=F,message=F,echo=F}
knitr::include_graphics("figBoxLine.png")
```

\newpage

```{r echo=F,warning=F,message=F,results='hide',fig.show='hide'}
generate_gene_hist<-function(data,label,xlimit){
    ggplot(data)+geom_histogram(aes(x=individuals),fill="grey",color="black",binwidth=500)+xlab("Number of Individuals")+ylab("Number of Genes")+theme_bw()+ggtitle(label)+xlim(-250,33000)
}
ddd_hist<-generate_gene_hist(ddd_all_merged,"DD",max(cardiac_all$individuals))
skin_hist<-generate_gene_hist(skin_all,"Skin",max(cardiac_all$individuals))
cancer_hist<-generate_gene_hist(cancer_all,"Cancer",max(cardiac_all$individuals))
eye_hist<-generate_gene_hist(eye_all,"Eye",max(cardiac_all$individuals))
cardiac_hist<-generate_gene_hist(cardiac_all,"Cardiac",max(cardiac_all$individuals))
CairoPNG("figGeneHist.png",width=1200,height=600)
grid.arrange(ddd_hist,skin_hist,cancer_hist,eye_hist,cardiac_hist,ncol=3,nrow=2)
dev.off()
CairoPNG("figGeneHist_single.png",width=1200,height=1500)
grid.arrange(ddd_hist,skin_hist,cancer_hist,eye_hist,cardiac_hist,ncol=1)
dev.off()
```
```{r geneHist, fig.cap="Histogram of number of individuals in UKB with variants within each gene in each panel. The width of each bin is 500 individuals.",echo=F,warning=F,message=F}
knitr::include_graphics("figGeneHist_single.png")
```

\newpage

```{r monoallelic, echo=F, warning=FALSE, message=F,results='hide',fig.show='hide'}
mod<-combined_panels_monoallelic %$% lm(individuals~length)
mono_labels<-combined_panels_monoallelic %>% filter((length<30000 & outlier==1)|(outlier==0 & individuals>5000 & length>12500 & length<20000) | (outlier==0 & length>12500 & individuals<4000 & length<30000))
mono_labels_1<-combined_panels_monoallelic %>% filter(length<30000 & outlier==0 & ((length>20000) | (length>12500 & length<14000 & individuals>4000 & individuals<5000)))
mono_labels_2<-combined_panels_monoallelic %>% filter(length<30000 & outlier==0 & length>14000 & length<20000 & individuals>4000 & individuals<5000)
text_size=2.5
mono_scatter<-combined_panels_monoallelic %>% filter(length<30000) %>% ggplot(aes(x=length,y=individuals,color=outlier,label=gene_name))+geom_point()+geom_abline(intercept=mod$coeff[1],slope=mod$coeff[2])+geom_text(data=mono_labels,nudge_y=-500,show.legend=F,size=text_size)+geom_text(data=mono_labels_1,nudge_x=-1700,show.legend=F,size=text_size)+geom_text(data=mono_labels_2,nudge_x=1700,show.legend=F,size=text_size)+theme_bw()+ggtitle("Monoallelic")+xlab("Coding Sequence Length (bp)")+ylab("Number of Individuals")+scale_colour_manual(values=cbPalette)
mono_scatter_panel<-combined_panels_monoallelic %>% filter(length<30000) %>% ggplot(aes(x=length,y=individuals,color=panel,label=gene_name))+geom_point()+geom_abline(intercept=mod$coeff[1],slope=mod$coeff[2])+geom_smooth(method="lm",show.legend=F)+geom_text(data=mono_labels,nudge_y=-500,show.legend=F,size=text_size)+geom_text(data=mono_labels_1,nudge_x=-1700,show.legend=F,size=text_size)+geom_text(data=mono_labels_2,nudge_x=1700,show.legend=F,size=text_size)+theme_bw()+ggtitle("Monoallelic")+xlab("Coding Sequence Length (bp)")+ylab("Number of Individuals")+scale_colour_manual(values=cbPalette)

mod<-combined_panels_biallelic %$% lm(individuals~poly(length,2,raw=T)) # generate model with length^2
x<-seq(0,max(combined_panels_biallelic$length),length.out=250)
y<-mod$coeff[1]+mod$coeff[2]*x+mod$coeff[3]*x^2
d2<-data.frame(cbind(x,y))
bi_labels<-combined_panels_biallelic %>% filter(outlier==1 & gene_name!="HSPG2" & gene_name!="HR" & gene_name!="KIAA1109") %>% distinct()
bi_labels_2<-combined_panels_biallelic %>% filter(outlier==1 & (gene_name=="HSPG2" | gene_name=="HR" | gene_name=="KIAA1109")) %>% distinct()
bi_labels_1<-combined_panels_biallelic %>% filter(length>20000) %>% distinct()
bi_scatter<-combined_panels_biallelic %>% ggplot(aes(x=length,y=individuals,color=outlier))+geom_point()+geom_line(data=d2,aes(x=x,y=y,color=NULL),show.legend=F)+geom_text(data=bi_labels,aes(label=gene_name),nudge_y=-120,show.legend=F,size=text_size)+geom_text(data=bi_labels_2,aes(label=gene_name),nudge_x=500,hjust=0,show.legend=F,size=text_size)+geom_text(data=bi_labels_1,aes(label=gene_name),nudge_y=-120,nudge_x=-750,show.legend=F,size=text_size)+theme_bw()+ggtitle("Biallelic")+xlab("Coding Sequence Length (bp)")+ylab("Number of Individuals")+theme(legend.position="None")+scale_colour_manual(values=cbPalette)
CairoPNG("outlier_scatter.png",width=1000,height=500)
grid.arrange(mono_scatter,bi_scatter,ncol=2)
dev.off()
# and create the full monoallelic figure for the Supplement
mod<-combined_panels_monoallelic %$% lm(individuals~length)
mono_labels<-combined_panels_monoallelic %>% filter(outlier==1)
mono_labels_1<-combined_panels_monoallelic %>% filter(outlier==0 & length>12500)
mono_scatter<-combined_panels_monoallelic %>% ggplot(aes(x=length,y=individuals,color=outlier,label=gene_name))+geom_point()+geom_abline(intercept=mod$coeff[1],slope=mod$coeff[2])+geom_text(data=mono_labels,nudge_y=-1000,show.legend=F)+geom_text(data=mono_labels_1,nudge_y=-1000,show.legend=F)+theme_bw()+ggtitle("Monoallelic")+xlab("Coding Sequence Length (bp)")+ylab("Number of Individuals")+scale_colour_manual(values=cbPalette)
CairoPNG("outlier_scatter_mono.png",width=1000,height=500)
mono_scatter
dev.off()
# create plot separating panels to see how concordant their regression lines are
bi_scatter<-combined_panels_biallelic %>% ggplot(aes(x=length,y=individuals,color=panel))+geom_point()+geom_line(data=d2,aes(x=x,y=y,color=NULL),show.legend=F)+geom_smooth(method="lm",show.legend=F,formula=y~poly(x,2,raw=T))+geom_text(data=bi_labels,aes(label=gene_name),nudge_y=-120,show.legend=F,size=text_size)+geom_text(data=bi_labels_2,aes(label=gene_name),nudge_x=500,hjust=0,show.legend=F,size=text_size)+geom_text(data=bi_labels_1,aes(label=gene_name),nudge_y=-120,nudge_x=-750,show.legend=F,size=text_size)+theme_bw()+ggtitle("Biallelic")+xlab("Coding Sequence Length (bp)")+ylab("Number of Individuals")+theme(legend.position="None")+scale_colour_manual(values=cbPalette)
CairoPNG("outlier_scatter_color_panel.png",width=1000,height=500)
grid.arrange(mono_scatter_panel,bi_scatter,ncol=2)
dev.off()
```
```{r outlierFig,fig.cap="Coding sequence length of gene vs number of individuals with at least 1 variant (monoallelic) or 2 variants (biallelic) in the gene. Lines of best-fit are linear for monoallelic and quadratic for biallelic genes. Genes further than 3 SD from the regression line are highlighted in blue. The x-axis of the monoallelic genes has been truncated to increase resolution. The full figure can be found in Figure SXX.", warning=F,message=F,echo=F}
knitr::include_graphics("outlier_scatter.png")
```

\newpage

```{r echo=F,warning=F,message=F,results='hide',fig.show='hide'}
revel<-read_tsv("results/DDD_revel_gene_variants_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic") %>% mutate(panel="DD")
revel<-read_tsv("results/Skin_revel_gene_variants_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic") %>% mutate(panel="Skin") %>% rbind(revel,.)
revel<-read_tsv("results/Eye_revel_gene_variants_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic") %>% mutate(panel="Eye") %>% rbind(revel,.)
revel<-read_tsv("results/Cancer_revel_gene_variants_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic") %>% mutate(panel="Cancer") %>% rbind(revel,.)
revel<-read_tsv("results/Cardiac_revel_gene_variants_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic") %>% mutate(panel="Cardiac") %>% rbind(revel,.)

mono1_revel<-revel %>% filter(inheritance=="monoallelic" & revel_quintile!=0) %>% ggplot(aes(x=length,y=individuals,color=as.factor(revel_quintile)))+geom_point()+geom_smooth(method="lm",show.legend=F)+xlab("Coding Sequence Length")+ylab("Number of Individuals")+ggtitle("Monoallelic")+theme_bw()+theme(legend.position="top",legend.text=element_text(size=20),legend.title=element_text(size=20))+guides(color=guide_legend(title="REVEL quintile"))+ylim(-500,max(revel$individuals)+200)+coord_cartesian(xlim=c(0,24000),expand=F)+scale_colour_manual(values=cbPalette)
mono2_revel<-revel %>% filter(inheritance=="monoallelic" & revel_quintile!=0) %>% ggplot(aes(x=length,y=individuals,color=as.factor(revel_quintile)))+geom_point()+geom_smooth(method="lm",show.legend=F)+xlab("")+ylab("")+theme_bw()+theme(legend.position="top")+theme(legend.title=element_blank(),legend.position="None")+ylim(-500,max(revel$individuals)+200)+theme(axis.text.y=element_blank(),axis.ticks.y=element_blank())+scale_x_continuous(breaks=c(108000))+coord_cartesian(xlim=c(107000,max(revel$length)+500),expand=F)+scale_colour_manual(values=cbPalette)
mono<-revel %>% filter(inheritance=="monoallelic" & !is.na(transcript_id) & revel_quintile!=0) %>% ggplot(aes(x=length,y=individuals,color=as.factor(revel_quintile)))+geom_point()+geom_smooth(method="lm",show.legend=F)+xlab("Coding Sequence Length")+ylab("Number of Individuals")+ggtitle("Monoallelic")+theme_bw()+theme(legend.position="top")+theme(legend.title=element_blank())+scale_x_break(c(24000,107000))+xlim(0,max(revel$length)+500)+ylim(0,max(revel$individuals)+200)+scale_colour_manual(values=cbPalette)
legend<-get_legend(mono1_revel)
mono1_revel<-mono1_revel+theme(legend.position="None")
bi_revel<-revel %>% filter(inheritance=="biallelic" & !is.na(transcript_id) & revel_quintile!=0) %>% ggplot(aes(x=length,y=individuals,color=as.factor(revel_quintile)))+geom_point()+geom_smooth(method="lm",show.legend=F,formula=y~poly(x,2,raw=T))+xlab("Coding Sequence Length")+ylab("Number of Individuals")+ggtitle("Biallelic")+theme_bw()+theme(legend.position="top")+guides(color=guide_legend(title="REVEL quintile"))+scale_colour_manual(values=cbPalette)
bi_revel<-bi_revel+theme(legend.position="None")
layout="
AAAAAAAAAABCCCCCCCCCCC
"
CairoPNG("revel_fig.png",width=1000,height=500)
mono1_revel+mono2_revel+bi_revel+plot_layout(design=layout)
dev.off()
```

```{r echo=F,warning=F,message=F,results='hide',fig.show='hide'}
# redefine cbPalette to make pathogenic red
cbPalette_clinvar <- c("#009E73", "#E69F00", "#D55E00", "#009E73", "#F0E442", "#0072B2", "#CC79A7")
# read in each dataset
ddd<-read_tsv("results/DDD_clinvar_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
skin<-read_tsv("results/Skin_clinvar_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
cancer<-read_tsv("results/Cancer_clinvar_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
eye<-read_tsv("results/Eye_clinvar_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
ddd_rdif<-read_tsv("results/DDD_RD_IF_clinvar_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
cardiac<-read_tsv("results/Cardiac_clinvar_gene_variant_counts",show_col_types = FALSE) %>% filter(inheritance=="biallelic" | inheritance=="monoallelic")
# join the DDD and DDD RD and IF genes
ddd_merged<-rbind(ddd,ddd_rdif) %>% mutate(id=paste(gene_name,inheritance,deleteriousness,sep="-"))
d2<-ddd_merged %>% select(gene_name,gene_id,transcript_id,inheritance,deleteriousness,id)
ddd_merged<-aggregate(ddd_merged[,6:9],list(ddd_merged$id),mean) %>% dplyr::rename(id=Group.1) %>% merge(.,d2) %>% select(gene_name,gene_id,transcript_id,inheritance,variants,individuals,length,total_length,deleteriousness)
rm(d2)
combined_panels<-ddd_merged %>% mutate(panel="DD")
combined_panels<-skin %>% mutate(panel="Skin") %>% rbind(combined_panels,.)
combined_panels<-eye %>% mutate(panel="Eye") %>% rbind(combined_panels,.)
combined_panels<-cancer %>% mutate(panel="Cancer") %>% rbind(combined_panels,.)
combined_panels<-cardiac %>% mutate(panel="Cardiac") %>% rbind(combined_panels,.)
combined_panels %<>% mutate(deleteriousness=ifelse(deleteriousness=="benign","Benign/Likely Benign",ifelse(deleteriousness=="pathogenic","Pathogenic/Likely Pathogenic",ifelse(deleteriousness=="conflicting","Conflicting interpretations",deleteriousness))))
combined_panels_monoallelic<-combined_panels %>% filter(inheritance=="monoallelic")
combined_panels_biallelic<-combined_panels %>% filter(inheritance=="biallelic")

## generate the plots
# remove uncertain
mono_scatter_clinvar<-combined_panels_monoallelic %>% filter(deleteriousness!="uncertain" & deleteriousness !="path-and-benign") %>% ggplot(aes(x=length,y=individuals,color=deleteriousness,label=gene_name))+geom_point()+geom_smooth(method="lm",show.legend=F)+theme_bw()+ggtitle("Monoallelic")+xlab("Coding Sequence Length")+ylab("Number of Individuals")+guides(color=guide_legend(title="ClinVar Classification"))+scale_colour_manual(values=cbPalette_clinvar)+coord_cartesian(xlim=c(0,24000),expand=F)+theme(legend.text=element_text(size=20),legend.title=element_text(size=20))
mono_scatter_clinvar_2<-combined_panels_monoallelic %>% filter(deleteriousness!="uncertain" & deleteriousness !="path-and-benign") %>% ggplot(aes(x=length,y=individuals,color=deleteriousness,label=gene_name))+geom_point()+geom_smooth(method="lm",show.legend=F)+theme_bw()+ggtitle("Monoallelic")+xlab("Coding Sequence Length")+ylab("Number of Individuals")+guides(color=guide_legend(title="ClinVar Classification"))+scale_colour_manual(values=cbPalette_clinvar)+coord_cartesian(xlim=c(107000,max(revel$length)+500),expand=F)+theme(legend.position="None")
mono2_revel<-revel %>% filter(inheritance=="monoallelic" & revel_quintile!=0) %>% ggplot(aes(x=length,y=individuals,color=as.factor(revel_quintile)))+geom_point()+geom_smooth(method="lm",show.legend=F)+xlab("")+ylab("")+theme_bw()+theme(legend.title=element_blank(),legend.position="None")+ylim(-500,max(revel$individuals)+200)+theme(axis.text.y=element_blank(),axis.ticks.y=element_blank())+scale_x_continuous(breaks=c(108000))+coord_cartesian(xlim=c(107000,max(revel$length)+500),expand=F)+scale_colour_manual(values=cbPalette)+theme(legend.position="None")
clinvar_legend<-get_legend(mono_scatter_clinvar)
mono_scatter_clinvar <- mono_scatter_clinvar+theme(legend.position="None")
bi_scatter_clinvar<-combined_panels_biallelic %>% filter(deleteriousness!="uncertain" & deleteriousness!="path-and-benign") %>% ggplot(aes(x=length,y=individuals,color=deleteriousness))+geom_point()+geom_smooth(method="lm",show.legend=F,formula=y~poly(x,2,raw=T))+theme_bw()+ggtitle("Biallelic")+xlab("Coding Sequence Length")+ylab("Number of Individuals")+theme(legend.position="None")+scale_colour_manual(values=cbPalette_clinvar)
CairoPNG("clinvar_fig_no_unc.png",width=1000,height=500)
grid.arrange(mono_scatter_clinvar,bi_scatter_clinvar,ncol=2)
dev.off()
CairoPNG("clinvar_fig_mono_no_unc.png",width=1000,height=500)
mono_scatter<-combined_panels_monoallelic %>% filter(deleteriousness!="uncertain" & deleteriousness!="path-and-benign") %>% ggplot(aes(x=length,y=individuals,color=deleteriousness,label=gene_name))+geom_point()+geom_smooth(method="lm",show.legend=F)+theme_bw()+ggtitle("Monoallelic")+xlab("Coding Sequence Length")+ylab("Number of Individuals")+scale_colour_manual(values=cbPalette_clinvar)
mono_scatter
dev.off()
layout="
AAAAAAAAAABCCCCCCCCCCD
AAAAAAAAAABCCCCCCCCCCD
AAAAAAAAAABCCCCCCCCCCD
AAAAAAAAAABCCCCCCCCCCD
EEEEEEEEEEEFFFFFFFFFFF
EEEEEEEEEEEFFFFFFFFFFF
EEEEEEEEEEEFFFFFFFFFFF
EEEEEEEEEEEFFFFFFFFFFF
GGGGGGGGGGHHHHHHHHHHHH
"
CairoPNG("revel_clinvar_fig.png",width=1200,height=1200)
mono1_revel+mono2_revel+mono_scatter_clinvar+mono_scatter_clinvar_2+bi_revel+bi_scatter_clinvar+legend+clinvar_legend+plot_layout(design=layout)
dev.off()
```
```{r clinvarFigNoUnc,fig.cap="Scatter plot of length of gene against number of individuals with variants identified in that gene, split by ClinVar annotation of the variant with the most sever consequence (Monoallelic) or the least severe consequence of the 2 highest consequence variants (Biallelic)",warning=F,message=F,echo=F}
knitr::include_graphics("revel_clinvar_fig.png")
```

\newpage

# Supplemental Figures {-}

```{r cancer_cases_cont_gene, echo=F,warning=F,message=F,results='hide',fig.show='hide'}
case_hist<-cancer_cases %>% ggplot()+geom_histogram(aes(x=individuals),fill="grey",color="black")+xlab("Number of Individuals")+ylab("Number of Genes")+theme_bw()+ggtitle("Cases")
cont_hist<-cancer_controls %>% ggplot()+geom_histogram(aes(x=individuals),fill="grey",color="black")+xlab("Number of Individuals")+ylab("Number of Genes")+theme_bw()+ggtitle("Controls")
CairoPNG("cancer_case_cont.png",height=500,width=1000)
grid.arrange(case_hist,cont_hist,ncol=2)
dev.off()
```
```{r cancerCases,fig.cap="Histograms of the  number of individuals with variants within each gene in the Cancer panel, split by cases and controls.",warning=F,message=F,echo=F}
knitr::include_graphics("cancer_case_cont.png")
```

\newpage

```{r outlierScatterMono, warning=F,message=F,echo=F,fig.cap="Coding sequence length of gene vs number of individuals with at least 1 variant in the gene. Genes further than 3 SD from the regression line are highlighted in blue."}
knitr::include_graphics("outlier_scatter_mono.png")
```

\newpage
```{r echo=F,warning=F,message=F,echo=F,fig.show='hide',results='hide'}
cbPalette_cis <- c("#D55E00", "#009E73", "#E69F00", "#009E73", "#F0E442", "#0072B2", "#CC79A7")
ddd_cis_trans<-read_tsv("DDD_cohort/results/DDD_genes_counts_cis_trans")
CairoPNG("cis_trans.png",width=1000,height=500)
ddd_cis_trans %<>% filter(!is.na(gene_id))
ggplot(data=ddd_cis_trans[which(ddd_cis_trans$deleteriousness!="unknown"),],aes(x=length,y=individuals,colour=deleteriousness))+geom_point(data=ddd_cis_trans[which(ddd_cis_trans$deleteriousness=="unknown"),])+geom_point(data=ddd_cis_trans[which(ddd_cis_trans$deleteriousness!="unknown"),])+geom_smooth(method="lm",show.legend=F,formula=y~poly(x,2,raw=T))+theme_bw()+theme(legend.title=element_blank())+scale_colour_manual(values=cbPalette_cis)+ylab("Number of Individuals")
dev.off()
```
```{r cisScatter,fig.cap="Scatter plot of length of gene against number of variants identified in the gene for biallelic genes for individuals in the DDD cohort, split by whether they appear in cis, trans, or undetermined based on parental genotype.",warning=F,message=F,echo=F}
knitr::include_graphics("cis_trans.png")
```

\newpage

```{r echo=F,warning=F,message=F,echo=F,fig.show='hide',results='hide'}
# Make Supplemental Tables
# ENS IDs
ens_id<-read_tsv("ensembl_gene_transcript_lengths") %>% select("Gene stable ID","Gene name") %>% distinct()
#DDD table
ddd_tab<-read_csv("g2p_files/DDG2P.csv") %>% select("gene symbol","allelic requirement") %>% rename("Gene Name"="gene symbol",moi="allelic requirement")
ddd_tab<-read_csv("g2p_files/DDD_RD_IFG2P.csv") %>% select("gene symbol","allelic requirement") %>% rename("Gene Name"="gene symbol",moi="allelic requirement") %>% rbind(ddd_tab,.) %>% distinct()
mois<-ddd_tab %>% select(moi) %>% distinct() %>% head(n=2)  %>% pull(moi) # get MOIs to keep - note not the most robust way of doing this but for some reason I can't get the strings to match if I manually specify them
ddd_tab %<>% filter(moi %in% mois) %>% rename("Mode of Inheritance"=moi)
# now add in ensembl IDs for the genes
ddd_tab <- ens_id %>% filter(`Gene name` %in% ddd_tab$`Gene Name`) %>% merge(ddd_tab,.,by.x="Gene Name",by.y="Gene name",all.x=T) %>% distinct()
# and merge with our list to exclude possible/probable and other oddities
ddd_tab <- combined_panels %>% filter(panel=="DD") %>% select(gene_name,gene_id,transcript_id,inheritance) %>% distinct() %>% merge(.,y=ddd_tab,by.y=c("Gene stable ID","Mode of Inheritance"),by.x=c("gene_id","inheritance"),all.x=T) %>% select("gene_name","inheritance","gene_id","transcript_id") %>% rename("Transcript stable ID"=transcript_id,"Gene Name"=gene_name,"Mode of Inheritance"=inheritance,"Gene stable ID"=gene_id) %>% distinct()
write.table(ddd_tab,"STableDDD.txt",quote=F,row.names=F,sep="\t")

cardiac_tab<-read_csv("g2p_files/CardiacG2P.csv") %>% select("gene symbol","allelic requirement") %>% rename("Gene Name"="gene symbol",moi="allelic requirement")
cardiac_tab %<>% filter(moi %in% c("biallelic","monoallelic")) %>% rename("Mode of Inheritance"=moi)
cardiac_tab <- ens_id %>% filter(`Gene name` %in% cardiac_tab$`Gene Name`) %>% merge(cardiac_tab,.,by.x="Gene Name",by.y="Gene name",all.x=T) %>% distinct()
cardiac_tab <- combined_panels %>% filter(panel=="Cardiac") %>% select(gene_name,gene_id,transcript_id,inheritance) %>% distinct() %>% merge(.,y=ddd_tab,by.y=c("Gene stable ID","Mode of Inheritance"),by.x=c("gene_id","inheritance"),all.x=T) %>% select("gene_name","inheritance","gene_id","transcript_id") %>% rename("Transcript stable ID"=transcript_id,"Gene Name"=gene_name,"Mode of Inheritance"=inheritance,"Gene stable ID"=gene_id) %>% distinct()
write.table(cardiac_tab,"STableCardiac.txt",quote=F,row.names=F,sep="\t")

cancer_tab<-read_csv("g2p_files/CancerG2P.csv") %>% select("gene symbol","allelic requirement") %>% rename("Gene Name"="gene symbol",moi="allelic requirement")
cancer_tab %<>% filter(moi %in% c("biallelic","monoallelic")) %>% rename("Mode of Inheritance"=moi)
cancer_tab <- ens_id %>% filter(`Gene name` %in% cancer_tab$`Gene Name`) %>% merge(cancer_tab,.,by.x="Gene Name",by.y="Gene name",all.x=T) %>% distinct()
cancer_tab <- combined_panels %>% filter(panel=="Cancer") %>% select(gene_name,gene_id,transcript_id,inheritance) %>% distinct() %>% merge(.,y=ddd_tab,by.y=c("Gene stable ID","Mode of Inheritance"),by.x=c("gene_id","inheritance"),all.x=T) %>% select("gene_name","inheritance","gene_id","transcript_id") %>% rename("Transcript stable ID"=transcript_id,"Gene Name"=gene_name,"Mode of Inheritance"=inheritance,"Gene stable ID"=gene_id) %>% distinct()
write.table(cancer_tab,"STableCancer.txt",quote=F,row.names=F,sep="\t")

eye_tab<-read_csv("g2p_files/EyeG2P.csv") %>% select("gene symbol","allelic requirement") %>% rename("Gene Name"="gene symbol",moi="allelic requirement")
eye_tab %<>% filter(moi %in% c("biallelic","monoallelic")) %>% rename("Mode of Inheritance"=moi)
eye_tab <- ens_id %>% filter(`Gene name` %in% eye_tab$`Gene Name`) %>% merge(eye_tab,.,by.x="Gene Name",by.y="Gene name",all.x=T) %>% distinct()
eye_tab <- combined_panels %>% filter(panel=="Eye") %>% select(gene_name,gene_id,transcript_id,inheritance) %>% distinct() %>% merge(.,y=ddd_tab,by.y=c("Gene stable ID","Mode of Inheritance"),by.x=c("gene_id","inheritance"),all.x=T) %>% select("gene_name","inheritance","gene_id","transcript_id") %>% rename("Transcript stable ID"=transcript_id,"Gene Name"=gene_name,"Mode of Inheritance"=inheritance,"Gene stable ID"=gene_id) %>% distinct()
write.table(eye_tab,"STableEye.txt",quote=F,row.names=F,sep="\t")

skin_tab<-read_csv("g2p_files/SkinG2P.csv") %>% select("gene symbol","allelic requirement") %>% rename("Gene Name"="gene symbol",moi="allelic requirement")
skin_tab %<>% filter(moi %in% c("biallelic","monoallelic")) %>% rename("Mode of Inheritance"=moi)
skin_tab <- ens_id %>% filter(`Gene name` %in% skin_tab$`Gene Name`) %>% merge(skin_tab,.,by.x="Gene Name",by.y="Gene name",all.x=T) %>% distinct()
skin_tab <- combined_panels %>% filter(panel=="Skin") %>% select(gene_name,gene_id,transcript_id,inheritance) %>% distinct() %>% merge(.,y=ddd_tab,by.y=c("Gene stable ID","Mode of Inheritance"),by.x=c("gene_id","inheritance"),all.x=T) %>% select("gene_name","inheritance","gene_id","transcript_id") %>% rename("Transcript stable ID"=transcript_id,"Gene Name"=gene_name,"Mode of Inheritance"=inheritance,"Gene stable ID"=gene_id) %>% distinct()
write.table(skin_tab,"STableSkin.txt",quote=F,row.names=F,sep="\t")
```
